Add --generate-api-docs to build script

Arcanemagus opened the pull request:
### Requirements

* Filling out the template is required. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* All new code requires tests to ensure against regressions

### Description of the Change

Add a new option `--generate-api-docs` that changes the build script to only do what is required to build the API documentation, skipping the process of building the full Atom binaries.

### Alternate Designs

Continuing to hack the build scripts manually to follow this behavior whenever a documentation only change is needed.

### Why Should This Be In Core?

It is an enhancement for the build script already in core.

### Benefits

Adds the ability to _only_ build the documentation when testing changes that only affect that and don't need the full binary to be built.

### Possible Drawbacks

Slightly higher complexity to the build script.

### Verification Process

Tested that this only does the necessary steps to build the documentation.

### Applicable Issues

None.

Arcanemagus committed:
Add --generate-api-docs to build script

Add a new option `--generate-api-docs` that changes the build script to
only do what is required to build the API documentation, skipping the
process of building the full Atom binaries.

daviwil commented on script/build:
```diff
@@ -21,11 +17,16 @@ const argv = yargs
   .wrap(yargs.terminalWidth())
   .argv
 
+// Needed so we can require src/module-cache.coffee during generateModuleCache
```
Moving this just because it isn't used in the previous section or was there another benefit?

Arcanemagus replied:
I think this is a leftover from a separate change where I was trying to get rid of _all_ the dependency `require`s (to exit early when `--help` is called), however as the arguments are parsed _in_ a dependency, it doesn't actually make that much of a difference. I can remove this change if you would prefer.

daviwil replied:
Removing it would be ideal, thanks a bunch!

daviwil commented on script/build:
```diff
@@ -72,7 +73,13 @@ if (!argv.existingBinaries) {
+}
+
+if (argv.generateApiDocs) {
+  process.exit()
```
I think you'll want to wait for the promise to complete before exiting, just in case.  Maybe it's better to add another conditional around the block of code below to check for the presence of that flag before adding the `.then()`s to `binariesPromise`?  That way the process will exit on its own after the promise finishes if we're only generating API docs.

Arcanemagus replied:
As at this stage `binariesPromise` is just a reference to `Promise.resolve()` that hasn't been called yet, there are no "in-flight" promises to wait for a resolution of as all of the above code is synchronous.

That being said, wrapping the entire block below in a check of that flag _is_ the more correct way of doing this if you want me to do that.

daviwil replied:
Yep, using the if statement might make it a bit more clear that the promise isn't just being ignored.  `binariesPromise` gets assigned from `dumpSymbols` if `--existing-binaries` isn't passed, so there will actually be a pending promise at that point.

daviwil reviewed (commented):
Thanks for adding the new argument, this will be handy! Just a couple comments here, nothing major

Arcanemagus committed:
Revert moving of dependencies

This change isn't necessary for adding the option.

Arcanemagus committed:
Wrap binariesPromise handling in an if

Instead of forcing an immediate exit with process.exit(), wrap the
handling of binariesPromise in an `if` statement, allowing the script to
exit on its own if passed the new option.

Arcanemagus commented:
@daviwil Updated, thanks for looking over this!

daviwil reviewed (approved):
Sweet, thanks a lot for making the changes!
